### ISSUE TITLE
Question: How to determine local drive path of UNC ie path C:\pathofshare from \\server\share

### ISSUE THREAD

**Issue by albvar** [alphaleonis/AlphaFS#200]
Simplified question with code examples:
`var driveInfo = new Alphaleonis.Win32.Filesystem.DriveInfo(@"\\\\server\sharename");` `driveInfo.VolumeInfo.Name` is returning the name of the volume label. I however do not see any properties that display the local directory ie C:\pathtoshare 

Question in layman's terms:
Trying to find the drive letter for a given unc path. For example I want to provide
\\server\sharename I want to see [DriveLetter]:\shares\sharename or something similar - the most important part here is knowing what drive letter the volume is located under. 

Thanks in advance

**Comment by Yomodo** [comment]
Unfortunately this is not possible, `DriveInfo` has no idea about/nor does it care, if the UNC path
points to a drive letter. (Being able to use a UNC path is one of the major differences compared to `System.IO`)

There exists a method: [Paths.LocalToUnc()](http://alphafs.alphaleonis.com/doc/2.0/index.html) which will convert a local path to a UNC path.

If you call this method with a mapped drive letter as parameter: `Path.LocalToUnc("V:\")`,
the UNC path of that drive letter will be returned.

In regard to the layman's question: This would require a method that checks each given UNC path to see if it is defined on the local host, someting like: `NET.EXE SHARE` command.

**Comment by brutaldev** [comment]
You can P/Invoke the the `NetShareEnum` method in `Netapi32.dll` which will give you a `NetShareData2` object that contains the local path. This technique even works on shares to remote machines.

**Comment by Yomodo** [comment]
@brutaldev Right on!
@albvar AlphaFS already has this, I forgot.

From my laptop, I enumerate the shares from my server: Yomodo

`Network.Host.EnumerateShares()` from host: [yomodo]

```
#001    Share: [\\yomodo\home]

Instance: [Alphaleonis.Win32.Network.ShareInfo]

#001    CurrentUses        = [0]
#002    DirectoryInfo      = [\\yomodo\home]
#003    MaxUses            = [4294967295]
#004    NetFullPath        = [\\yomodo\home]
#005    NetName            = [home]
#006    Password           = []
#007    Path               = [C:\var\services\homes\admin]
#008    Permissions        = [None]
#009    Remark             = [home]
#010    ResourceType       = [None]
#011    SecurityDescriptor = [0]
#012    ServerName         = [yomodo]
#013    ShareLevel         = [Info2]
#014    ShareType          = [DiskTree]
```

**Comment by brutaldev** [comment]
Quick look at the network classes in AlphaFS and this is already possible for a single path (since AplhaFS is using `NetShareEnum` behind the scenes).

``` csharp
var shareInfo = Alphaleonis.Win32.Network.Host.GetShareInfo(@"\\machine\sharename", true);
var localPath = shareInfo.Path;
```

**Comment by Yomodo** [comment]
Yep :+1:

**Comment by albvar** [comment]
Maybe I asked the wrong question - I am not looking to find the fully qualified share location, instead 1. I am looking to find the volume path for my share. Say I have a share that lives in C:\temp\shares as \myserver\shares$ I want the volume path to return C:
2. Lets assume that my share lives in C:\temp\mount\InternationalShares  where mount is a mounted disk and my share unc is \myservers\internationalShares$. I want to return C:\temp\mount as this is where the disk was mounted.

@everyone thanks for helping me.
@brutaldev --> referencing your latest comment...for some reason shareInfo comes up null
var shareInfo = Alphaleonis.Win32.Network.Host.GetShareInfo(@"\myserver\myshare", true);
var enumShare = new List<ShareInfo>(Host.EnumerateShares("myserver", true));
//shareInfo is null?

enumShare[5].DirectoryIfno = \myserver\myshare
enumShare[5].Path = "VolPath:\subfldr\share"   //not what I am interested in - want to see the volume path where the share is.

**Comment by brutaldev** [comment]
@albvar Missing a slash in your network path there, needs to start with a double-slash...

Still don't quite understand your new explanation, if you just want the drive letter behind the share just use the `Path` methods.

``` csharp
// Will return "C:" for example.
Alphaleonis.Win32.Filesystem.Path.GetPathRoot(shareInfo.Path).TrimEnd(Alphaleonis.Win32.Filesystem.Path.DirectorySeparatorChar);
```

If you're looking to determine the underlying path of a mount, you need to make more P/Invokes again, this time [GetVolumePathName](https://msdn.microsoft.com/en-us/library/windows/desktop/aa364996%28v=vs.85%29.aspx) or [GetVolumeNameForVolumeMountPoint ](https://msdn.microsoft.com/en-us/library/windows/desktop/aa364994%28v=vs.85%29.aspx) will get what you're looking for depending on your scenario.

AlphaFS does provide wrappers for these methods via `Alphaleonis.Win32.Filesystem.Volume`

**Comment by albvar** [comment]
@brutaldev -> thanks once again - for some reason when I posted my comment it stripped out slashes. I am using the correct parameter. var shareInfo = Alphaleonis.Win32.Network.Host.GetShareInfo(@"\\myserver\myshare", true);

Regarding not understanding: I have a scenario where I have mounted volumes and the VolumeInfo wrapper doesn't seem to be mount aware. For a share that is on a mounted volume at C:\mountedVolume\share it'll return C:\ when I need c:\mountedVolume. I will look at the API's you have referenced - thanks for recommendation.

Although I was not able to find a solution to my problem using this library I was able to figure out a different way which is to query Win32_Volume which is a WMI class after figuring out what server the UNC path is being hosted on. In an environment that uses DFS and Cluster resource names for unc paths this is a "easier said than done" applies here. I will be unable to provide examples because of this.

**Comment by Yomodo** [comment]
> For a share that is on a mounted volume at C:\mountedVolume\share it'll return C:\ when I need c:\mountedVolume.

I think this should do the trick?
[GetVolumePathName](https://msdn.microsoft.com/en-us/library/windows/desktop/aa364996(v=vs.85).aspx) function:

> If a specified path is passed, GetVolumePathName returns the path to the volume mount point, which means that it returns the root of the volume where the end point of the specified path is located.

> For example, assume that you have volume D mounted at C:\Mnt\Ddrive and volume E mounted at "C:\Mnt\Ddrive\Mnt\Edrive". Also assume that you have a file with the path "E:\Dir\Subdir\MyFile". If you pass "C:\Mnt\Ddrive\Mnt\Edrive\Dir\Subdir\MyFile" to GetVolumePathName, it returns the path "C:\Mnt\Ddrive\Mnt\Edrive\".
